[FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)aims to provide a simple way to do some quality control checks on raw sequence data coming from high throughput sequencing pipelines. It provides a modular set of analyses which you can use to give a quick impression of whether your data has any problems of which you should be aware before doing any further analysis.

The main functions of FastQC are

Import of data from BAM, SAM or FastQ files (any variant)
Providing a quick overview to tell you in which areas there may be problems
Summary graphs and tables to quickly assess your data
Export of results to an HTML based permanent report
Offline operation to allow automated generation of reports without running the interactive application

# Code

Inside the folder with samples 
```ruby
module load fastqc

fastqc *.fastq.gz
fastqc *.fastq
```


# Results

For majority of F1

```
PASS	Basic Statistics	
PASS	Per base sequence quality	
PASS	Per tile sequence quality	
PASS	Per sequence quality scores	
FAIL	Per base sequence content	
WARN	Per sequence GC content	
PASS	Per base N content	
PASS	Sequence Length Distribution	
FAIL	Sequence Duplication Levels
WARN	Overrepresented sequences	
WARN	Adapter Content	
```

Clem mentioned human contamination in the samples
