In order to get haplotypes from both parental lines we require the transcriptomes from both. In order to get the transcriptome from Kaz, we use the parental samples, map them and get the SNPs that later will be converted to a transcriptome of its own.

## Indexing

Indexing allows to determine the position and orientation information for where the fragments best map prior to quantification. The reference index essentially provides the transcriptome in a format that is easily and rapidly searchable


Change Asinica500.fa for the .fa file of your parental transcriptome

``` ruby
module load bwa
srun bwa index Asinica500.fa
```

## Map reads to transcriptome

![image](https://github.com/sarabi98/howToImprinting/assets/94226596/dbf66044-d731-43a7-b81b-5a92fc4ec848)


Use the next command:

srun bwa mem -M -t 30 transcriptome.fa sample_1.fastq.gz sample_2.fastq.gz | samtools view -b | samtools sort -T individual > sample_sort.bam

Changing transcriptome.fa for the parental transcriptome file and sample_1.fastq.gz, sample_2.fastq.gz for both paired end files per sample.

``` ruby
module load samtools
module load bwa
#srun bwa mem -M -t 30 transcriptome.fa sample_1.fastq.gz sample_2.fastq.gz | samtools view -b | samtools sort -T individual > sample_sort.bam
srun bwa mem -M -t 30 Asinica500.fa /nfs/scistore18/vicosgrp/madamega/artemiaImprinting/analysis/parents/97661_P_KAZ_F.1.fastq.gz /nfs/scistore18/vicosgrp/madamega/artemiaImprinting/analysis/parents/97661_P_KAZ_F.2.fastq.gz | samtools view -b | samtools sort -T individual > 97661_sort.bam
srun bwa mem -M -t 30 Asinica500.fa /nfs/scistore18/vicosgrp/madamega/artemiaImprinting/analysis/parents/39873_P_KAZ_M.1.fastq.gz /nfs/scistore18/vicosgrp/madamega/artemiaImprinting/analysis/parents/39873_P_KAZ_M.2.fastq.gz | samtools view -b | samtools sort -T individual > 39873_sort.bam
srun bwa mem -M -t 30 Asinica500.fa /nfs/scistore18/vicosgrp/madamega/artemiaImprinting/analysis/parents/39874_P_KAZ_M.1.fastq.gz /nfs/scistore18/vicosgrp/madamega/artemiaImprinting/analysis/parents/39874_P_KAZ_M.2.fastq.gz | samtools view -b | samtools sort -T individual > 39874_sort.bam
srun bwa mem -M -t 30 Asinica500.fa /nfs/scistore18/vicosgrp/madamega/artemiaImprinting/analysis/parents/39875_P_KAZ_F.1.fastq.gz /nfs/scistore18/vicosgrp/madamega/artemiaImprinting/analysis/parents/39875_P_KAZ_F.2.fastq.gz | samtools view -b | samtools sort -T individual > 39875_sort.bam
srun bwa mem -M -t 30 Asinica500.fa /nfs/scistore18/vicosgrp/madamega/artemiaImprinting/analysis/parents/39876_P_KAZ_F.1.fastq.gz /nfs/scistore18/vicosgrp/madamega/artemiaImprinting/analysis/parents/39876_P_KAZ_F.2.fastq.gz | samtools view -b | samtools sort -T individual > 39876_sort.bam
```

# SNP calling

srun bcftools mpileup -a AD,DP,SP -Ou -f transcriptome.fa sample.bam ... sample.bam  | bcftools call -v -f GQ,GP -mO z -o head.vcf.gz

Notice the  filter  parameters: -minQ --minDP --maxDP

They change the amount of transcripts that can be detected by the parental counts

srun vcftools --gzvcf head.vcf.gz  --remove-indels --minQ 30 --minDP 5 --maxDP 500 --recode --stdout >  head_filtered.vcf

bcftools view --max-alleles 2 --exclude-types indels head_filtered.vcf > transcriptome2.vcf

``` ruby
module load samtools
module load bcftools
module load vcftools

#first index the transcriptome
srun samtools faidx Asinica500.fa

srun bcftools mpileup -a AD,DP,SP -Ou -f Asinica500.fa 97661_sort.bam 39873_sort.bam 39874_sort.bam 39875_sort.bam 39876_sort.bam | bcftools call -v -f GQ,GP -mO z -o head.vcf.gz

#filter for quality and coverage
srun vcftools --gzvcf head.vcf.gz  --remove-indels --minQ 30 --minDP 5 --maxDP 500 --recode --stdout >  head_filteredKDP.vcf

#Filter 2: remove multiallelic
bcftools view --max-alleles 2 --exclude-types indels head_filteredKDP.vcf > /nfs/scistore18/vicosgrp/madamega/artemiaImprinting/analysis/VCF_DP_better/kazWholeBody.vcf
```

# VCF to .fa

Create a consensus sequence for an individual where the sequence incorporates typed variants from the reads.


gzip -k transcriptome2.vcf
cat transcriptome2.vcf | grep -v "DEL" | grep -v "INS" > transcriptome2.vcf
bgzip -c transcriptome2.vcf > transcriptome2.vcf.gz
tabix -p vcf transcriptome2.vcf.gz

## 3.2. reconstruct the transcriptome
cat transcriptome.fa | bcftools consensus transcriptome2.vcf.gz > transcriptome2.fa

``` ruby
module load bcftools

#3.1. Remove INDELS and index the files
gzip -k kazWholeBody.vcf
cat kazWholeBody.vcf | grep -v "DEL" | grep -v "INS" > nkazWholeBody.vcf
bgzip -c nkazWholeBody.vcf > nkazWholeBody.vcf.gz
tabix -p vcf nkazWholeBody.vcf.gz

## 3.2. reconstruct the genomes
cat Asinica500.fa | bcftools consensus nkazWholeBody.vcf.gz > kazPWB_SNPs.genrec.fa
```
