From this point on run in python.

Note: The functions are made with my files as a basis, the function cleanFile must be changed after the comment "right ID and parental labels" to fit the ase-tigar output id and parental descriptions if other files are used. Other functions reference "Sin", "Kaz", "C1KS" or "C2SK" if using other parental types these taggs should also be changed!

Don't forget to import all necessary libraries:

``` python
import pandas as pd
import glob
import numpy as np
from matplotlib import pyplot as plt
import random
import os
import math
import seaborn as sns
```
## Create table with all the data

### Funtions

Clean file: Takes the path to the folder with the ASE-TIGAR output files and returns a dataframe with refine ID name and parental info.
``` python
def cleanFile(path_nameFile):
    df=pd.read_csv(path_nameFile,delimiter="\t")
    #right ID and parental labels 
    df['ID'] = df['ID'].str.replace('/nfs/scistore18/vicosgrp/madamega/artemiaImprinting/analysis/VCF_DP_better/','')
    df['ID'] = df['ID'].str.replace('PWB_SNPs.genrec_ArtemiaSinica','')
    df[['Parental', 'ID']] = df['ID'].str.split('_', n=1, expand=True)
    df=df.rename(columns={"ID":"Gene"}) #Gene -> transcript
    clean=df
    return clean
```

getTable: Turns input path(s) to folder(s) of ASE-TIGAR outputs and returns dataframe in table form with all the clean data, calls funciton cleanFile. 
```python
def getTable(*paths):
    data=pd.DataFrame()
    for path in paths:
        for filename in glob.iglob(path+'**/*ASE.txt', recursive = True):
            size=len(filename.split("/"))
            name=filename.split("/")[size-1]
            name=name.split("_ASE.txt")[0]
            name=name.split("_")
            id=name[0]
            type=name[2]
            sex=name[3]
            #call funtion to clean gene id info 
            df=cleanFile(filename)
            #asign info of sample
            df["ID"]=id
            df["Type"]=type
            df["Sex"]=sex
            #reorder
            df=df[['ID','Type','Sex','Gene','Parental','LENGTH','Z','FPKM','THETA']]
            data=data.append(df)
    return data
```
### How does the table look:


![image](https://github.com/sarabi98/howToImprinting/assets/94226596/fa1a6f9b-acb3-4dd9-a44c-dab7cec4ca6a)

### Running the code

paths to output files from ase-tigar
```python
pathC1="artemiaImprinting/analysis/version3_betterVCF/outputC1/"
pathC2="artemiaImprinting/analysis/version3_betterVCF/outputC2/"
kaz='artemiaImprinting/analysis/version3_betterVCF/outputSC/kaz/'
sin='artemiaImprinting/analysis/version3_betterVCF/outputSC/sin/'
data=getTable(pathC1,pathC2,kaz,sin)
data
```

Save table as .txt file 

```python
data.to_csv('artemiaImprinting/data.txt',index=False,sep="\t")
```

## Obtaining proportion values kaz/(kaz+sin) 

### Functions

pValues: Inputs a dataframe, sex (speficy with a string F for females, M for males, keep empty or B for both), test (boolean values True if you want to make a random sampling test with the data nC1 and nC2 must be specified or False or empty if all the samples must be taken which is the normal way to use it), nc1 (int value, number of samples of C1 to take at random, keep empty if test==False), nc2 (int value, number of samples of C2 to take at random, keep empty if test==False). 

Uses a pivot table to sum all the values in Z that have the same Gene value (same transcript) per Parental type, then it adds a p column calculating kaz/(kaz+sin) and returns the new dataframe called p.

```python
def pValues(df,sex='B',test=False,nC1=None,nC2=None):
    if sex!='B':
        dataF=df[df['Sex']==sex]
    else:
        dataF=df
    if test==True:
        #it most return both parental trasncripts for the sample id
        tc1=df[df['Type']=='C1KS'].sample(nC1)
        tc1=tc1['ID'].tolist()
        tc2=df[df['Type']=='C2SK'].sample(nC2)
        tc2=tc2['ID'].tolist()
        tc=tc1+tc2 #list of samples selected
        dataF=df[df['ID'].isin(tc)]
    p=pd.DataFrame()
    p= pd.pivot_table(dataF, values='Z', index=['Gene'],columns=['Parental'],aggfunc='sum')
    p["P"]=p["kaz"]/(p["kaz"]+p["sin"])
    return p
```

plotHistP: Plots histograms of proportion values, x axes from 0 to 1. Takes p dataframe as input and a string with the description of the dataframe.

```python
def plotHistP(data,name):
    data['P'].plot(kind="hist",bins=50,title="Histogram P values "+name)
    plt.xlabel("P=kaz/(kaz+sin)")
    plt.ylabel("Frequency")
    plt.xlim([0, 1]) 
```

plotHistR: Plot histoogram of mat/pat ratio with x axes limits of -1 to 1. Takes the data with the ratio, a name string with the description and a boolean lim False or empty if there is no y axe limits and True if it is limited from 0 to 100.

```python
def plotHistR(data,name,lim=False):
    plt.hist(data, bins=50, alpha=0.5)
    plt.title("Histogram Mat/Pat P2-P1 ratio "+name)
    plt.xlabel("Mat/Pat expression")
    plt.ylabel("Frequency")
    plt.xlim([-1, 1]) 
    if lim:
        plt.ylim([0,100])
```

plotExpressionLog: Takes a previously filtered dataframe only containing the information of one transcript, a name string with the descritption, sex string with F for females M for males, empty or B for both and a log boolean value True if the graph will be in log scale log2(x+1) or False if its in Z counts scale.

```python
def plotExpressionLog(gene, name, sex='B',log=True):
    if sex != 'B':
        gene = gene[gene['Sex'] == sex]
    
    if log==True:
        title='Levels of expression: Counts of parental haplotypes log2(x+1) for: '
    else:
        title='Levels of expression: Counts of parental haplotypes for: '
        
    fig, axes = plt.subplots(2, 2, figsize=(40, 10), dpi=100, facecolor='w', edgecolor='k')
    fig.suptitle(title + name)

    types = ['C1KS', 'C2SK', 'SIN', 'KAZ']
    colors = ['blue', 'orange']

    for i, t in enumerate(types):
        current_ax = axes[i // 2, i % 2]

        subset = gene[gene['Type'] == t]
        if log==True:
            subset['Z'] = [math.log2(x + 1) for x in subset['Z']]
        subset = subset.pivot(index=['ID','Sex'], columns='Parental', values='Z')
        subset = subset.sort_values('Sex')
        print(subset)
        subset.plot(ax=current_ax, kind='bar', color=[colors[1], colors[0]])
        current_ax.get_legend().remove() #so sin and kaz legend only appears in the figure
        current_ax.set_title(t)
    # Create a common legend outside subplots
    fig.legend(['Kaz', 'Sin'], loc='upper right')

    plt.tight_layout(rect=[0, 0, 1, 1])
    plt.show()
```
### Using data to run the code

```python
C1=data[data['Type']=='C1KS']
C2=data[data['Type']=='C2SK']
sin=data[data['Type']=='SIN']
kaz=data[data['Type']=='KAZ']

#getting p values 
C1p=pValues(C1)
C2p=pValues(C2)
sinp=pValues(sin)
kazp=pValues(kaz)

#specific depending on sex
C1pF=pValues(C1,'F')
C2pF=pValues(C2,'F')

C1pM=pValues(C1,'M')
C2pM=pValues(C2,'M')

plotHistP(sinp,"parental sin")
plotHistP(kazp,"parental kaz")
plotHistP(C1p,"C1")
plotHistP(C2p,"C2")
```
![image](https://github.com/ClemLasne/SHARED_SINKAZ/assets/94226596/29cfedd2-2092-422f-b7c3-c111b374b10d)
![image](https://github.com/ClemLasne/SHARED_SINKAZ/assets/94226596/f7379ac7-5a8e-4950-b21c-88920bf30c6c)
![image](https://github.com/ClemLasne/SHARED_SINKAZ/assets/94226596/d51489f7-e6c5-4431-ab0b-dadce870437e)
![image](https://github.com/ClemLasne/SHARED_SINKAZ/assets/94226596/fbab2b54-b6d8-431d-a92f-deaee2bb9901)

Genes recognized by parental lines

```python
#recognized by parents
recIDkaz=kazp[kazp["P"]>= 0.8]
print(len(recIDkaz['P']),'Number of genes recognized by kaz parents')
recIDsin=sinp[sinp["P"]<= 0.2]
print(len(recIDsin['P']),'Number of genes recognized by sin parents')
rec=recIDkaz.index[recIDkaz.index.isin(recIDsin.index)].tolist()
print(len(rec),'Number of genes recognized by both parental lines')

>12126 Number of genes recognized by kaz parents
>12964 Number of genes recognized by sin parents
>9444 Number of genes recognized by both parental lines

#filtering out recognized by parents from p datasets
C1p=C1p[C1p.index.isin(rec)]
C2p=C2p[C2p.index.isin(rec)]

C1pF=C1pF[C1pF.index.isin(rec)]
C2pF=C2pF[C2pF.index.isin(rec)]

C1pM=C1pM[C1pM.index.isin(rec)]
C2pM=C2pM[C2pM.index.isin(rec)]
```
New plots, recognized by parents

```python
plotHistP(C1p,"C1 KxS")
plotHistP(C2p,"C2 SxK")
```
![image](https://github.com/ClemLasne/SHARED_SINKAZ/assets/94226596/ec8f7716-7d8b-4df4-a194-beddf4400ea7)
![image](https://github.com/ClemLasne/SHARED_SINKAZ/assets/94226596/6f087e0f-b114-4bb4-bf45-261156e58566)

Getting ratios and ploting 

```python
ratio=C2p['P']-C1p['P']
plotHistR(ratio,"F1")
candidates= [x for x in ratio if x < -0.3 or x > 0.3]
plotHistR(candidates,"fitering out x>0.3",True)
print(len(candidates))
> 414

#getting The candidates
can=ratio.loc[(ratio<-0.75) | (ratio>0.75)]
can=can.sort_values()
can

#females
females=C2pF['P']-C1pF['P']
plotHistR(females,'females')
canF= [x for x in females if x < -0.3 or x > 0.3]
plotHistR(canF,"Females fitering out x>0.3",True)
print(len(canF))
>925

#males
males=C2pM['P']-C1pM['P']
plotHistR(males,'Males')
canM= [x for x in males if x < -0.3 or x > 0.3]
plotHistR(canM,"Males fitering out x>0.3",True)
print(len(canM))
>677
```
![image](https://github.com/ClemLasne/SHARED_SINKAZ/assets/94226596/7c68b26f-f825-4183-a317-e26e3ef1c395)
![image](https://github.com/ClemLasne/SHARED_SINKAZ/assets/94226596/a237a862-f579-44fd-ad7b-1befe04760ec)

Females

![image](https://github.com/ClemLasne/SHARED_SINKAZ/assets/94226596/012b4d79-a02a-409a-a947-663fa0e6e352)
![image](https://github.com/ClemLasne/SHARED_SINKAZ/assets/94226596/5adc4ac3-f695-4d2e-9875-f1a48688a2a1)

Males

![image](https://github.com/ClemLasne/SHARED_SINKAZ/assets/94226596/147ba6ed-3099-4796-a3a5-cd3f6007cf4b)
![image](https://github.com/ClemLasne/SHARED_SINKAZ/assets/94226596/ada0292e-c505-4a76-83d3-94024a92fd8e)

### The candidates


```python
print(can)
>

24 genes
Gene
A11.R23369365                -1.000000
d12777741412742              -1.000000
A13.R17680514                -0.999940
d1463350857414               -0.999275
A19.R7680939                 -0.945554
TRINITY_DN311019_c3_g1_i1    -0.943604
A18.R8954563                 -0.917500
A5.R46748560                 -0.907830
TRINITY_DN309529_c1_g1_i12   -0.883838
A19.J7734245                 -0.874302
A2.S39767445                 -0.857143
A18.S9156277                 -0.810901
A7.R39607491                 -0.795988
A18.R8900166                 -0.751645
TRINITY_DN299500_c1_g1_i1     0.764385
A8.R34187460                  0.774194
A19.R7696037                  0.844244
A15.S13526733                 0.952381
A7.R39585895                  0.996000
A16.S11908149                 0.999541
A12.S20044720                 1.000000
A19.R7674919                  1.000000
A5.R47408230                  1.000000
A4.R54128716                  1.000000
```


Filtering for a specific transcript
```python
#filtering
gene=data[data["Gene"]=='A13.R17680514'] #the gene that always shows up


```
The candidates
```python
	ID	Type	Sex	Gene	Parental	LENGTH	Z	FPKM	THETA
7327	231319	C1KS	M	A13.R17680514	sin	624	0.00	0.000000	0.000000e+00
28215	231319	C1KS	M	A13.R17680514	kaz	624	71598.00	10361.175667	1.695652e-03
7327	231303	C1KS	F	A13.R17680514	sin	624	2.00	0.328206	5.025052e-08
28215	231303	C1KS	F	A13.R17680514	kaz	624	97122.89	15938.167243	2.440238e-03
7327	231324	C1KS	F	A13.R17680514	sin	624	2.00	0.238692	3.382947e-08
28215	231324	C1KS	F	A13.R17680514	kaz	624	179378.00	21408.054410	3.034131e-03
7327	231323	C1KS	F	A13.R17680514	sin	624	2.00	0.250413	3.907791e-08
28215	231323	C1KS	F	A13.R17680514	kaz	624	152383.99	19079.470849	2.977424e-03
7327	231302	C1KS	F	A13.R17680514	sin	624	3.00	0.842807	6.141191e-08
28215	231302	C1KS	F	A13.R17680514	kaz	624	84139.89	23637.897119	1.722397e-03
7327	231318	C1KS	M	A13.R17680514	sin	624	3.00	0.346100	5.102549e-08
28215	231318	C1KS	M	A13.R17680514	kaz	624	142736.15	16466.984907	2.427727e-03
7327	231304	C1KS	F	A13.R17680514	sin	624	0.00	0.000000	4.492549e-40
28215	231304	C1KS	F	A13.R17680514	kaz	624	123445.82	18531.061194	2.723139e-03
7327	231310	C1KS	M	A13.R17680514	sin	624	0.00	0.000000	7.006492e-45
28215	231310	C1KS	M	A13.R17680514	kaz	624	167685.36	15375.854088	2.775433e-03
7327	231308	C1KS	M	A13.R17680514	sin	624	0.00	0.000000	1.933999e-36
28215	231308	C1KS	M	A13.R17680514	kaz	624	158358.94	22548.758461	3.535932e-03
7327	231317	C1KS	M	A13.R17680514	sin	624	1.00	0.201156	2.143962e-08
28215	231317	C1KS	M	A13.R17680514	kaz	624	75438.01	15174.806062	1.617362e-03
7327	231309	C1KS	M	A13.R17680514	sin	624	10.00	2.786128	2.683892e-07
28215	231309	C1KS	M	A13.R17680514	kaz	624	81192.18	22621.176857	2.179111e-03
7327	231305	C2SK	F	A13.R17680514	kaz	624	0.00	0.000000	5.363807e-30
28215	231305	C2SK	F	A13.R17680514	sin	624	142065.18	31584.543861	3.085944e-03
7327	231314	C2SK	M	A13.R17680514	kaz	624	4.00	0.680613	7.000317e-08
28215	231314	C2SK	M	A13.R17680514	sin	624	171330.00	29152.366506	2.998410e-03
7327	231312	C2SK	M	A13.R17680514	kaz	624	0.00	0.000000	2.223193e-36
28215	231312	C2SK	M	A13.R17680514	sin	624	114981.00	19031.638909	2.349013e-03
7327	231315	C2SK	M	A13.R17680514	kaz	624	18.00	3.076788	3.777353e-07
28215	231315	C2SK	M	A13.R17680514	sin	624	93165.00	15924.941178	1.955095e-03
7327	231313	C2SK	M	A13.R17680514	kaz	624	2.00	0.272919	3.416462e-08
28215	231313	C2SK	M	A13.R17680514	sin	624	119592.22	16319.485174	2.042911e-03
7327	231316	C2SK	M	A13.R17680514	kaz	624	6.00	0.824693	1.102989e-07
28215	231316	C2SK	M	A13.R17680514	sin	624	144741.01	19894.488606	2.660795e-03
7327	231307	C2SK	F	A13.R17680514	kaz	624	4.00	0.603143	8.550949e-08
28215	231307	C2SK	F	A13.R17680514	sin	624	126626.58	19093.479000	2.706944e-03
7327	231306	C2SK	F	A13.R17680514	kaz	624	15.00	3.103032	3.017956e-07
28215	231306	C2SK	F	A13.R17680514	sin	624	119199.00	24658.556025	2.398249e-03
7327	231311	C2SK	M	A13.R17680514	kaz	624	0.00	0.000000	6.614269e-36
28215	231311	C2SK	M	A13.R17680514	sin	624	114561.00	20668.719318	2.281302e-03
7327	39875	KAZ	F	A13.R17680514	sin	624	25.00	6.250176	1.563170e-06
28215	39875	KAZ	F	A13.R17680514	kaz	624	103148.85	25787.937199	6.449565e-03
7327	39873	KAZ	M	A13.R17680514	sin	624	11.00	2.385506	5.524848e-07
28215	39873	KAZ	M	A13.R17680514	kaz	624	111618.75	24213.421951	5.607845e-03
7327	97661	KAZ	F	A13.R17680514	sin	624	0.00	0.000000	2.841660e-38
28215	97661	KAZ	F	A13.R17680514	kaz	624	283180.80	26230.867483	7.856454e-03
7327	39876	KAZ	F	A13.R17680514	sin	624	36.00	25.403296	4.406627e-06
28215	39876	KAZ	F	A13.R17680514	kaz	624	58867.50	41539.677933	7.205753e-03
7327	39874	KAZ	M	A13.R17680514	sin	624	18.00	5.779887	1.034141e-06
28215	39874	KAZ	M	A13.R17680514	kaz	624	130904.37	42034.023935	7.520752e-03
7327	39870	SIN	M	A13.R17680514	sin	624	212176.30	34146.593948	7.614832e-03
28215	39870	SIN	M	A13.R17680514	kaz	624	1.00	0.160935	3.588918e-08
7327	39869	SIN	M	A13.R17680514	sin	624	173032.22	21500.588851	5.828506e-03
28215	39869	SIN	M	A13.R17680514	kaz	624	12.96	1.610487	4.365803e-07
7327	39872	SIN	F	A13.R17680514	sin	624	79134.26	10778.783084	3.273799e-03
28215	39872	SIN	F	A13.R17680514	kaz	624	27.58	3.756375	1.140909e-06
7327	39871	SIN	F	A13.R17680514	sin	624	35978.51	11866.050418	2.932767e-03
28215	39871	SIN	F	A13.R17680514	kaz	624	6.00	1.978857	4.890866e-07
```


### Extra Random Sampling

```python
ran1=pValues(data,test=True,nC1=5,nC2=6)
ran2=pValues(data,test=True,nC1=7,nC2=6)
ran1=ran1[ran1.index.isin(rec)]
ran2=ran2[ran2.index.isin(rec)]
ratRan=ran2['P']-ran1['P']
plotHistR(ratRan,"random samples")
candiRan=[x for x in ratRan if x < -0.3 or x > 0.3]
plotHistR(candiRan,"fitering out x>0.3",True)
print(len(candiRan))
>146

canRan=ratRan.loc[(ratRan<-0.75) | (ratRan>0.75)]
canRan=canRan.sort_values()
canRan

```
![image](https://github.com/ClemLasne/SHARED_SINKAZ/assets/94226596/5e7499a7-0e25-4590-9b76-b50bc192b584)
![image](https://github.com/ClemLasne/SHARED_SINKAZ/assets/94226596/18ea109a-785a-4edc-a92f-da9f8dd6e0ed)


